runtime/rococo/src/propose_parachain.rs

```diff
@@ -0,0 +1,836 @@
+		/// Proposals that are approved.
+		ApprovedProposals: Vec<ParaId>;
+		/// Proposals that are scheduled at for a fixed session to be applied.
+		ScheduledProposals: map hasher(twox_64_concat) SessionIndex => Vec<ParaId>;
+		/// Information about the registered parachains.
+		ParachainInfo: map hasher(twox_64_concat) ParaId => Option<RegisteredParachainInfo<T::AccountId, T::ValidatorId>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin, system = system {
+		type Error = Error<T>;
+
+		/// The maximum name length of a parachain.
+		const MaxNameLength: u32 = T::MaxNameLength::get();
+
+		fn deposit_event() = default;
+
+		#[weight = T::MaximumBlockWeight::get()]
+		fn propose_parachain(
+			origin,
+			para_id: ParaId,
+			name: Vec<u8>,
+			validation_function: ValidationCode,
+			initial_head_state: HeadData,
+			validators: [T::ValidatorId; 2],
+			balance: BalanceOf<T>,
+		) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(name.len() <= T::MaxNameLength::get() as usize, Error::<T>::NameTooLong);
+			ensure!(!Proposals::<T>::contains_key(&para_id), Error::<T>::ParachainIdAlreadyProposed);
+			ensure!(!runtime_common::parachains::Code::contains_key(&para_id), Error::<T>::ParachainIdAlreadyTaken);
+
+			let active_validators = Session::<T>::validators();
+			ensure!(
+				active_validators.iter().all(|v| *v != validators[0] && *v != validators[1]),
+				Error::<T>::ValidatorAlreadyRegistered,
+			);
+			Proposals::<T>::iter().try_for_each(|(_, prop)|
+				if prop.validators.iter().all(|v| *v != validators[0] && *v != validators[1]) {
+					Ok(())
+				} else {
+					Err(Error::<T>::ValidatorAlreadyRegistered)
+				}
+			)?;
+
+			T::Currency::reserve(&who, T::ProposeDeposit::get())?;
+
+			let proposal = Proposal {
+				name: name.clone(),
+				proposer: who,
+				validators: validators.into(),
+				initial_head_state,
+				validation_function,
+				balance,
+			};
+
+			Proposals::<T>::insert(para_id, proposal);
+
+			Self::deposit_event(Event::ParachainProposed(name, para_id));
+		}
+
+		#[weight = 100_000]
+		fn approve_proposal(
+			origin,
+			para_id: ParaId,
+		) {
+			T::PriviledgedOrigin::try_origin(origin).map_err(|_| BadOrigin)?;
+
+			ApprovedProposals::append(para_id);
+
+			Self::deposit_event(Event::ParachainApproved(para_id));
+		}
+
+		#[weight = 100_000]
+		fn cancel_proposal(origin, para_id: ParaId) {
+			let who = match EnsurePriviledgedOrSigned::<T>::try_origin(origin).map_err(|_| BadOrigin)? {
+				Either::Left(()) => None,
+				Either::Right(who) => Some(who),
+			};
+
+			let proposal = Proposals::<T>::get(&para_id).ok_or(Error::<T>::ProposalNotFound)?;
+
+			if let Some(who) = who {
+				ensure!(who == proposal.proposer, Error::<T>::NotAuthorized);
+			}
+
+			Proposals::<T>::remove(&para_id);
+
+			T::Currency::unreserve(&proposal.proposer, T::ProposeDeposit::get());
```

Comment:
Depending on who has access to Rococo, you should consider slashing the deposit here if called by the privileged origin. (assuming the proposal was rejected in that case)

Otherwise, there is no real disincentive to just ask for a million proposals... But maybe this kind of thinking only applies to a production chain

Reply:
My current idea is to require a high deposit to prevent spam xD And the tokens will come from us :P

So, yeah, I'm ignoring this all here.